KoPrivateGPT/benchmark/retrival/retrieval_evaluation_master.py

```diff
@@ -0,0 +1,91 @@
+from base_retrieval_evaluation_factory import RecallFactory, PrecisionFactory, NDCGFactory, MAPFactory, MRRFactory, \
+    HoleFactory, TopKAccuracyFactory #RecallCapFactory
+
+# from benchmark.retrival.f1 import F1Factory
+# from benchmark.retrival.average_precision import AveragePrecisionFactory
+
+# from benchmark.retrival.pfound import PFoundFactory
+
+from typing import List, Dict, Union, Tuple
+
+import json
+import click
+
+
+def retrieval_evaluation_master(qrels: Dict[str, Dict[str, int]],
+                                results: Dict[str, Dict[str, float]],
+                                k_values: List[int]) -> List[dict[str, float]]:
+    """
+    This function is the master function for the retrival evaluation.
+    It calls the other functions and returns a dictionary with the metrics as keys and the values as values.
+    :param qrels: The qrels file as a dictionary. Dict[query_id, Dict[doc_id, relevance]]
+    :param results: The results file as a dictionary. Dict[query_id, Dict[doc_id, score]]
+    results doc_id can be different from the doc_id in the qrels file.
+    """
+    evaluation_results = []
+    metrics_factory = [RecallFactory, PrecisionFactory, NDCGFactory, MAPFactory, MRRFactory, HoleFactory,
+                       TopKAccuracyFactory]#RecallCapFactory
+
+    max_length = max(len(retrieved_docs) for retrieved_docs in results.values())
+    filtered_k_values, _ = [k for k in k_values if (max_length >= k >= 1)], [k for k in k_values if
```

Comment:
Plus, you can just use .filter method for this.

Reply:
...! I didn't 'Python' at all.

Reply:
variable _ is gone...
then I got a sexier code than before...!